blog/_posts/2019-08-20-multi-container-live-update.md

```diff
@@ -0,0 +1,77 @@
+---
+slug: multi-container-live-update
+date: 2019-08-20
+author: maia
+layout: blog
+title: "Live Update Multiple Containers Per Pod"
+subtitle: "Because your sidecar deserves iteration too!"
+image: sidecarDog.jpg
+image_caption: "Who WOULDN'T want to iterate on this sidecar?!"
+tags:
+  - docker
+  - kubernetes
+  - microservices
+  - tilt
+  - containers
+keywords:
+  - live-update
+  - terminal
+  - tilt
+  - development
+---
+Back in the Dark Ages (i.e. last week), the lion’s share of Tilt’s logic relied on the assumption that there was only one container per pod that was worth caring about. This was the container that we got restart counts from, that we monitored for stale code, and that we updated in place. If you were running your code on _more than one_ container, well, we could only Live Update[^1] the first one, because that was the only one we knew about. Any changes to the other containers would result in a full `docker build && kubectl apply`.
+
+But, as the Bard said: "There are more workflows in heaven and Earth, Horatio, than are dreamt of in your initial product spec." (Or something.) So, we got cracking on some new functionality: as of [Tilt v0.10.0](https://github.com/windmilleng/tilt/releases), you can use Live Update on as many containers per pod as you want! This change didn’t change much on the surface of Tilt, but involved a bunch of restructuring and rewriting under the hood--we had to change a bunch of internal assumptions to account for multiple multiple per pod. This touched everything from logs (`kubectl get logs podabc -c mycontainer`) to resource readiness checks to health monitoring--but we think it was worth it!
+
+[Here’s an example repo](https://github.com/windmilleng/sidecar_example) you can take for a spin, based on the premise that you might want to iterate on your app and your home-rolled sidecar at the same time. In this Kubernetes YAML, you can see that we have two different containers running on a single pod:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+...
+spec:
+  ...
+    spec:
+      containers:
+        - name: randword
+          image: randword
+          ...
+        - name: log-ingester
+          image: log-ingester
+```
+
+The app is called `randword`, and every second, it logs a random word to a file. The log ingester runs the same pod as the app, reads in the app's logfile by means of a shared volume, and "ingests" the logs (here, just printing them to stdout).
+
+The Tilfile seems simple, because it’s probably what you expected you could write all along, right up until the point you tried, and got the error: "Tilt only supports Live Updates for the first Tilt-built container on a pod." But now it works!
```

Comment:
*Tiltfile

Reply:
geez matt, it's the today-i-learned file!